flux-kube/plugins/flux-kube.c

```diff
@@ -0,0 +1,324 @@
+    flux_kvs_txn_t *txn = NULL;
+    flux_jobid_t *idptr = NULL;
+
+    if (!(h = flux_jobtap_get_flux (p))
+        || flux_job_kvs_key (key, sizeof (key), id, "R") < 0
+        || !(txn = flux_kvs_txn_create ())
+        || flux_kvs_txn_put (txn, 0, key, R) < 0
+        || !(f = flux_kvs_commit (h, NULL, 0, txn))
+        || flux_future_then (f, -1, alloc_continuation, p)
+        || !(idptr = calloc (1, sizeof (*idptr)))
+        || flux_future_aux_set (f, "jobid", idptr, free) < 0) {
+        flux_kvs_txn_destroy (txn);
+        flux_future_destroy (f);
+        free (idptr);
+        return -1;
+    }
+    *idptr = id;
+    flux_kvs_txn_destroy (txn);
+    return 0;
+}
+
+
+static int sched_cb (flux_plugin_t *p,
+                     const char *topic,
+                     flux_plugin_arg_t *args,
+                     void *arg)
+{
+    const char *R, *jobspec = NULL;
+    flux_jobid_t id;
+    flux_future_t *f;
+    int retval = -1;
+
+    /*  If flux-kube::R set on this job then commit R to KVS,
+     *  get the jobspec, submit the jobspec to flux-kube.py, 
+     *  and set flux-kube flag.
+     */
+    if (!(R = flux_jobtap_job_aux_get (p,
+                                       FLUX_JOBTAP_CURRENT_JOB,
+                                       "flux-kube::R")))
+        return 0;
+
+    if (!(jobspec = flux_jobtap_job_aux_get (p,
+                                       FLUX_JOBTAP_CURRENT_JOB,
+                                       "flux-kube::jobspec")))
+        return 0;
+
+    if (flux_plugin_arg_unpack (args,
+                                FLUX_PLUGIN_ARG_IN,
+                                "{s:I}",
+                                "id", &id) < 0) {
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "flux-kube: %s: unpack: %s",
+                                     topic,
+                                     flux_plugin_arg_strerror (args));
+        return -1;
+    }
+
+    if (!(f = flux_rpc_pack (flux_jobtap_get_flux (p), "fluxkube.submit", 
+                             FLUX_NODEID_ANY, 0,
+                             "{s:s}", "jobspec", jobspec))) { 
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "flux-kube: %s: Failed to send RPC: %s",
+                                     topic,
+                                     strerror (errno));
+        return -1;
+    }
+    
+    if (flux_rpc_get_unpack (f, "{s:i}", "retval", &retval) != 0) {
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "flux-kube: %s: Failed to "
+                                     "unpack RPC: %s",
+                                     topic,
+                                     strerror (errno));
+        return -1;
+    }
```

Comment:
This should be moved to an asynchronous callback, otherwise this synchronous get will block the entire job manager. (i.e. use `flux_future_then(3)`) If there are other synchronous gets, those should be modified as well.

Reply:
What happens if `flux_rpc_get_unpack` blocks for a while (like 10 seconds)? Is there a timeout after which the call fails?  I ask because I may have run into this already.

Reply:
There is no timeout for `flux_rpc_get()` or `flux_rpc_get_unpack()`. The code will block and the job manager will be stuck permanently.

When switching to `flux_future_then(3)` you can optionally add a timeout in the second parameter. (The RPC future will be fulfilled with an error)

Reply:
This is probably a dumb question, but can I just wrap a `flux_rpc_get_unpack()` in a callback which I pass to `flux_future_then()`?

Reply:
If it's more involved (which is probably the case) we can chat during a coffee time.

Reply:
I think so. 

@milroy. Since you are familiar with the high-level resource API within fluxion. Perhaps comparing these two blocking vs. nonblocking functions would be useful?


https://github.com/flux-framework/flux-sched/blob/master/resource/hlapi/bindings/c%2B%2B/reapi_module_impl.hpp#L125

https://github.com/flux-framework/flux-sched/blob/master/resource/hlapi/bindings/c%2B%2B/reapi_module_impl.hpp#L55

Reply:
Yes, that answers my question. The nonblocking callback wraps a blocking `flux_rpc_get_unpack()` with some additional logic, so it's not as complex as I figured.

Reply:
One other comment. With Flux RPCs you can use `flux_respond_error()` to send an error response. Using that mechanism rather than encoding `retval` in the payload would possibly allow you to simplify the callback. (You can use `if (flux_rpc_get (f, NULL) < 0)` to test for error)

Reply:
The daemon is written in Python; Is there a Python equivalent to `flux_respond_error`?

Reply:
Yeah, the bindings are FFI so it _should_ work as `flux_handle.respond_error(msg, errno, "message")`

If you supply an error message, it can be decoded on the other side with `flux_future_error_string(3)`. There is a convenience macro `flux_future_strerror()` which will use an error string if available and if not will fall back to `strerror (errno)`

Reply:
Implementing `flux_handle.respond_error(msg, errno, "message")` doesn't seem to be working, but maybe that's because I'm passing `None` as the message string.

Reply:
How exactly is it not working?

FWIW, I verified that this works:

```python
#!/usr/bin/env python3
import flux
import errno

def cb(h, t, msg, arg):
    h.respond_error(msg, errno.EINVAL, "Testing respond error")

h = flux.Flux()
h.service_register("test").get()
w = h.msg_watcher_create (cb, flux.constants.FLUX_MSGTYPE_REQUEST, "test.*")
w.start()
h.reactor_run()
```

```console
$ flux mini submit flux python test.py
ƒ294fWJUdM
$ ./t/request/rpc test.foo </dev/null
rpc: test.foo: Testing respond error
```

Reply:
I also tested with:
```python
    h.respond_error(msg, errno.EINVAL, None)
```
and it worked for me:
```console
$ ./t/request/rpc test.foo </dev/null
rpc: test.foo: Invalid argument
```

Reply:
It wasn't working in that the error code I passed was never meeting the 
```C 
if (flux_rpc_get (f, NULL) < 0
``` 
condition.  That may be because I was using an `int` instead of `errno`

Reply:
I can get `respond_error` to work, provided I use a positive integer for success (`respond_error` fails for `0`). Is that abusing the functionality to use `respond_error` for success and failure?

Reply:
Alternatively, I can just use `respond_error` only when OpenShift fails to create the service, but then there could be many futures waiting forever for responses.

Reply:
Yeah, sorry I should have been more clear. `flux_respond_error()` creates an error response message so is only appropriate if there was an error. If there was success you can just use `handle.respond(msg)`

Reply:
Great! That works.